packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,489 @@
+        const assetsForChain = assetsByScope[chainId as CaipChainId];
+        // Now fetch metadata from the associated asset Snaps:
+        const snap = this.#getAssetSnapFor(chainId as CaipChainId);
+        if (snap) {
+          const metadata = await this.#getMetadata(assetsForChain, snap.id);
+          newMetadata = {
+            ...newMetadata,
+            ...(metadata ? metadata.assets : {}),
+          };
+        }
+      }
+    }
+    this.update((state) => {
+      state.metadata = {
+        ...this.state.metadata,
+        ...newMetadata,
+      };
+    });
+  }
+
+  /**
+   * Creates a mapping of CAIP-2 Chain ID to Asset Snaps.
+   *
+   * @returns A mapping of CAIP-2 Chain ID to Asset Snaps.
+   */
+  #getAssetSnaps(): Record<CaipChainId, Snap[]> {
+    const snaps: Record<CaipChainId, Snap[]> = {};
+    const allSnaps = this.#getAllSnaps();
+    const allPermissions = allSnaps.map((snap) =>
+      this.#getSnapsPermissions(snap.id),
+    );
+
+    for (const [index, permission] of allPermissions.entries() as unknown as [
+      number,
+      AssetEndowment,
+    ][]) {
+      const chainIdsCaveatPermission = permission[
+        'endowment:assets'
+      ]?.caveats.find((value) => value.type === 'chainIds');
+      const scopes = chainIdsCaveatPermission?.value;
+      if (!scopes) {
+        continue;
+      }
+      for (const scope of scopes as CaipChainId[]) {
+        if (!snaps[scope]) {
+          snaps[scope] = [];
+        }
+        snaps[scope].push(allSnaps[index]);
+      }
+    }
+    return snaps;
+  }
+
+  /**
+   * Returns the first asset snap for the given scope
+   *
+   * @param scope - The scope to get the asset snap for
+   * @returns The asset snap for the given scope
+   */
+  #getAssetSnapFor(scope: CaipChainId): Snap | undefined {
+    const allSnaps = this.#snaps[scope];
+    // Pick only the first one, we ignore the other Snaps if there are multiple candidates for now.
+    return allSnaps?.[0]; // Will be undefined if there's no Snaps candidate for this scope.
+  }
+
+  /**
+   * Returns all the asset snaps
+   *
+   * @returns All the asset snaps
+   */
+  #getAllSnaps(): Snap[] {
+    return this.messagingSystem
+      .call('SnapController:getAll')
+      .filter((snap) => snap.enabled && !snap.blocked);
+  }
+
+  /**
+   * Returns the permissions for the given origin
+   *
+   * @param origin - The origin to get the permissions for
+   * @returns The permissions for the given origin
+   */
+  #getSnapsPermissions(
+    origin: string,
+  ): SubjectPermissions<PermissionConstraint> {
+    return this.messagingSystem.call(
+      'PermissionController:getPermissions',
+      origin,
+    ) as SubjectPermissions<PermissionConstraint>;
+  }
+
+  /**
+   * Returns the metadata for the given assets
+   *
+   * @param assets - The assets to get metadata for
+   * @param snapId - The snap ID to get metadata from
+   * @returns The metadata for the assets
+   */
+  async #getMetadata(
```

Comment:
What about renaming this to `getAssetsMetadataFrom`? We already use `refreshAssetsMetadata` so I feel like having an `Assets` prefix is better here?

Reply:
Sounds good!